velox/type/Filter.cpp

```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  auto otherHugeintRange = dynamic_cast<const HugeintRange*>(&other);
+  return otherHugeintRange != nullptr && Filter::testingEquals(other) &&
+      lower_ == otherHugeintRange->lower_ &&
+      upper_ == otherHugeintRange->upper_;
+}
+
+folly::dynamic BigintValuesUsingHashTable::serialize() const {
+  auto obj = Filter::serializeBase("BigintValuesUsingHashTable");
+  obj["min"] = min_;
+  obj["max"] = max_;
+
+  folly::dynamic values = folly::dynamic::array;
+  for (auto v : values_) {
+    values.push_back(v);
+  }
+  obj["values"] = values;
+
+  folly::dynamic hashTable = folly::dynamic::array;
+  for (auto h : hashTable_) {
+    hashTable.push_back(h);
+  }
+  obj["hashTable"] = hashTable;
+
+  return obj;
+}
+
+FilterPtr BigintValuesUsingHashTable::create(const folly::dynamic& obj) {
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  auto min = obj["min"].asInt();
+  auto max = obj["max"].asInt();
+
+  std::vector<int64_t> values;
+  auto valArr = obj["values"];
+  values.reserve(valArr.size());
+  for (const auto& v : valArr) {
+    values.push_back(v.asInt());
+  }
+
+  std::vector<int64_t> hashTable;
+  auto hashArr = obj["hashTable"];
+  hashTable.reserve(hashArr.size());
+  for (const auto& h : hashArr) {
+    hashTable.push_back(h.asInt());
+  }
+  auto res = std::make_unique<BigintValuesUsingHashTable>(
+      min, max, values, nullAllowed);
+
+  return res;
+}
+
+bool BigintValuesUsingHashTable::testingEquals(const Filter& other) const {
+  auto otherBigintValuesUsingHashTable =
+      dynamic_cast<const BigintValuesUsingHashTable*>(&other);
+  bool res = otherBigintValuesUsingHashTable != nullptr &&
+      Filter::testingEquals(other) &&
+      min_ == otherBigintValuesUsingHashTable->min_ &&
+      max_ == otherBigintValuesUsingHashTable->max_ &&
+      containsEmptyMarker_ ==
+          otherBigintValuesUsingHashTable->containsEmptyMarker_ &&
+      sizeMask_ == otherBigintValuesUsingHashTable->sizeMask_ &&
+      hashTable_.size() == otherBigintValuesUsingHashTable->hashTable_.size() &&
+      values_.size() == otherBigintValuesUsingHashTable->values_.size();
+
+  if (!res) {
+    return false;
+  }
+
+  for (size_t i = 0; i < values_.size(); ++i) {
+    if (values_.at(i) != otherBigintValuesUsingHashTable->values_.at(i)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+folly::dynamic BigintValuesUsingBitmask::serialize() const {
+  auto obj = Filter::serializeBase("BigintValuesUsingBitmask");
+  obj["min"] = min_;
+  obj["max"] = max_;
+
+  folly::dynamic values = folly::dynamic::array;
+  for (size_t i = 0; i < bitmask_.size(); ++i) {
+    if (bitmask_[i]) {
+      values.push_back(static_cast<int64_t>(i) + min_);
+    }
+  }
+  obj["values"] = values;
+
+  return obj;
+}
+
+FilterPtr BigintValuesUsingBitmask::create(const folly::dynamic& obj) {
+  auto min = obj["min"].asInt();
+  auto max = obj["max"].asInt();
+  auto nullAllowed = obj["nullAllowed"].asBool();
+
+  std::vector<int64_t> values;
```

Comment:
This code seems to be repeating. Consider adding helper function: deserializeValues(obj)